File: backend/scripts/tokenBucket.lua
-- KEYS[1] = bucket:<userId>
-- ARGV[1] = BUCKET_SIZE
-- ARGV[2] = REFILL_RATE (tokens per second)

local bucketKey = KEYS[1]
local bucketSize = tonumber(ARGV[1])
local refillRate = tonumber(ARGV[2])

-- Get current Redis time (milliseconds)
local redisTime = redis.call("TIME")
local now =
    tonumber(redisTime[1]) * 1000 +
    math.floor(tonumber(redisTime[2]) / 1000)

-- Read bucket
local bucketData = redis.call("HGETALL", bucketKey)

local bucket = {}

-- Create bucket if it doesn't exist
if #bucketData == 0 then
    bucket.tokens = bucketSize
    bucket.lastRefill = now
else
    -- Convert Redis array into a Lua table
    for i = 1, #bucketData, 2 do
        bucket[bucketData[i]] = tonumber(bucketData[i + 1])
    end
end

-- Refill tokens
local elapsedTime = (now - bucket.lastRefill) / 1000
local tokensToAdd = elapsedTime * refillRate

bucket.tokens = math.min(
    bucketSize,
    bucket.tokens + tokensToAdd
)

-- Preserve fractional refill time
bucket.lastRefill =
    bucket.lastRefill +
    (tokensToAdd / refillRate) * 1000

-- Check if request is allowed
if bucket.tokens < 1 then
    redis.call(
        "HSET",
        bucketKey,
        "tokens", bucket.tokens,
        "lastRefill", bucket.lastRefill
    )

    return 0
end

-- Consume one token
bucket.tokens = bucket.tokens - 1

-- Save updated bucket
redis.call(
    "HSET",
    bucketKey,
    "tokens", bucket.tokens,
    "lastRefill", bucket.lastRefill
)

return 1

File: backend/service/tokenBucket.js
const BUCKET_SIZE = 1; // Maximum number of tokens in the bucket
const BUCKET_REFILL_RATE = 1; // Number of tokens added to the bucket per second

const buckets = new Map(); // Map to store token buckets for each user

export function allowRequest(userId) {
  if (!buckets.has(userId)) {
    buckets.set(userId, {
      tokens: BUCKET_SIZE,
      lastRefill: Date.now(),
    });
  }

  const bucket = buckets.get(userId);

  const currentTime = Date.now();

  const timeSinceLastFill = (currentTime - bucket.lastRefill) / 1000; // Convert to seconds

  const tokensToAdd = Math.floor(timeSinceLastFill * BUCKET_REFILL_RATE);

  // Refill bucket only if at least one token should be added
  if (tokensToAdd > 0) {
    bucket.tokens = Math.min(BUCKET_SIZE, bucket.tokens + tokensToAdd);

    bucket.lastRefill += (tokensToAdd / BUCKET_REFILL_RATE) * 1000;
  }

  if (bucket.tokens >= 1) {
    buckets.get(userId).tokens--;
    console.log(
      `Request allowed for user ${userId}. Remaining tokens: ${bucket.tokens}`,
    );
    return true;
  }
  return false;
}
